test/e2e/kubectl.go

```diff
@@ -0,0 +1,215 @@
+	//
+	// You can read about the syntax here: http://golang.org/pkg/text/template/
+	getContainerStateTemplate := fmt.Sprintf(`--template={{and (exists . "currentState" "info" "%s" "state" "running")}}`, container)
+
+	getImageTemplate := fmt.Sprintf(`--template={{(index .currentState.info "%s").image}}`, container)
+
+	getHostIPTemplate := "--template={{.currentState.hostIP}}"
+
+	By(fmt.Sprintf("waiting for all containers in %s pods to come up.", selector))
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(5 * time.Second) {
+		cmd := runKubectl(
+			"get", "pods", "-o", "template",
+			getPodsTemplate,
+			"-l", selector)
+		pods := strings.Fields(cmd.stdout.String())
+		if numPods := len(pods); numPods != replicas {
+			By(fmt.Sprintf("Replicas for %s: expected=%d actual=%d", selector, replicas, numPods))
+			continue
+		}
+		var runningPods []string
+		for _, podId := range pods {
+			running := runKubectl(
+				"get", "pods", podId, "-o", "template",
+				getContainerStateTemplate).stdout.String()
+			if running == "false" {
+				By(fmt.Sprintf("%s is created but not running", podId))
+				continue
+			}
+
+			currentImage := runKubectl(
+				"get", "pods", podId, "-o", "template",
+				getImageTemplate).stdout.String()
+			if currentImage != image {
+				By(fmt.Sprintf("%s is created but running wrong image; expected: %s, actual: %s", podId, image, currentImage))
+				continue
+			}
+
+			hostIP := runKubectl(
+				"get", "pods", podId, "-o", "template",
+				getHostIPTemplate).stdout.String()
+			data, err := getData(hostIP)
+			if err != nil {
+				By(fmt.Sprintf("%s is running right image but fetching data failed: %v", podId, err))
+				continue
+			}
+			if strings.Contains(data.image, image) {
+				By(fmt.Sprintf("%s is running right image but fetched data has the wrong info: %s", podId, data))
+				continue
+			}
+
+			By(fmt.Sprintf("%s is verified up and running", podId))
+			runningPods = append(runningPods, podId)
+		}
+		if len(runningPods) == replicas {
+			return
+		}
+	}
+	Fail(fmt.Sprintf("Timed out waiting for %s pods to reach valid state", selector))
+}
+
+type updateDemoData struct {
+	image string `json:"image"`
+}
+
+func getData(hostIP string) (*updateDemoData, error) {
+	addr := fmt.Sprintf("http://%s:8080/data.json", hostIP)
+	resp, err := http.Get(fmt.Sprintf(addr))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	By(fmt.Sprintf("got data: %s", body))
+	var data updateDemoData
+	err = json.Unmarshal(body, &data)
+	return &data, err
+}
+
+type kubectlCmd struct {
+	*exec.Cmd
+	stdout, stderr bytes.Buffer
+}
+
+func runKubectl(args ...string) *kubectlCmd {
+	path := filepath.Join(root, "cluster/kubectl.sh")
+	cmdStr := path + " " + strings.Join(args, " ")
+	By(fmt.Sprintf("Running '%v'", cmdStr))
+
+	cmd := &kubectlCmd{
```

Comment:
What is the advantage of making a struct here rather than just creating a *exec.Cmd and assigning Stdout/Stderr to a var on the local stack?

var stdout, stderr bytes.Buffer
cmd := exec.Command(path, args...)
cmd.Stdout, cmd.Stderr = stdout, stderr